### ISSUE TITLE
Filter by object tags

### ISSUE THREAD

**Issue by aidan-** [datacenter/acitoolkit#203]
This isn't necessarily an issue but more of a question, so please let me know if there is a more suitable place to ask this.

I'm essentially trying to subscribe to all EPG events and filter the EPG's by a tag associated to them.  Looking at the `subscribe` and `get_event` methods, the returned objects do not have their internal _tags list populated.  I did a little bit of digging around and it seems the only place the _tags list gets populated is using the `get_deep` method which does not have much documentation or usage in the samples or test cases.

Is this method the right way to do this, or am I missing something?  

TLDR; Knowing the DN of an EPG, what is the easiest way to retrieve that object from an APIC as well as tags associated to it.

**Comment by michsmit99** [comment]
Hi Aidan,
When you subscribe to class events i.e. EPG.subscribe(), you will only get event notifications for the EPG object. You could use this as a trigger to issue a get_deep call and get the EPG along with its tags.  For performance reasons, you could limit this get_deep to only EPGs and tags if desired through the get_deep limit_to parameter. Another option is to craft a specific query directly to the APIC API and use the Session.subscribe() function directly. This would require you to process the raw APIC response.  Perhaps, we can introduce new APIs to collect the tags in addition to object events. What type of EPG events were you looking for ?

**Comment by aidan-** [comment]
Hi Michael,
Thinking a little more about the problem, I think what I am ideally after is tag add/delete events as tags can be added/removed from an object after it has been created.

What I am working towards is writing a simple utility that automates the configuration of the _intersite_ application using tags to nominate whether an EPG should be replicated or not.

I will have a go at using the _Session.subscribe()_ function directly to see if that gives me what I want.  If this were to become a native method of the acitoolkit would this best be done as a super class to the _BaseACIObject_ (much like Tenant, EPG, etc) or is that too overkill?  

Do you have any examples of using the _get_deep_ method from ACIBaseObject (for EPG)?  Reading through the code, I'm not sure what some of the parameters are expecting or if the _limit_to_ variable will be effective.

**Comment by michsmit99** [comment]
Hi Aidan,
I just made a check in that will allow subscribing to Tags https://github.com/datacenter/acitoolkit/commit/9adc30b4d8eada95acab460939da72ad02cea4a2. It works in the same manner as the Tenant subscription example in https://github.com/datacenter/acitoolkit/blob/master/samples/aci-subscribe-tenants.py.  (Just replace all aci.Tenant references to aci.Tag)

I think we need to write a separate detailed FAQ for get_deep(). The limit_to is largely for performance optimization. It narrows the query sent to the APIC to the specified APIC object classes e.g. if you only wanted EPGs, you could set the limit_to to only include 'fvTenant', 'fvAp', and 'fvAEPg'. This would avoid pulling all of the Contracts, etc. from the APIC. Typically, such an optimization is not necessary.

**Comment by aidan-** [comment]
Hey Michael,
Finally getting some time to continue looking at this and it looks like it will work quite well! Thanks for your help.

I've come across a small issue where a Tag's parent object type isn't defined causing an error.  I've submitted a small PR with some more information in #212 .

**Comment by aidan-** [comment]
This looks to be working properly now.

Thanks again for the assistance!
